### PR TITLE
Feature/cuda height map

### PR DESCRIPTION
**Pull request by LaughingCommit**
added height map CUDA code

### CODE REVIEW

File: ihmc-high-level-behaviors/src/main/java/us/ihmc/PerceptionAndAutonomyProcess.java
Comment (TomaszTB) [inline ihmc-high-level-behaviors/src/main/java/us/ihmc/PerceptionAndAutonomyProcess.java:1]: You seem to have forgotten a few changes here

Come to think of it, I think I made these changes while we were testing something.. sorry!

File: ihmc-high-level-behaviors/src/main/java/us/ihmc/perception/RapidHeightMapManager.java [hunk 1]
```diff
@@ -29,13 +31,14 @@
...
-   private final RapidHeightMapExtractor rapidHeightMapExtractor;
+   private final RapidHeightMapExtractorCuda rapidHeightMapExtractor;
    private final ImageMessage croppedHeightMapImageMessage = new ImageMessage();
    private final FramePose3D cameraPoseForHeightMap = new FramePose3D();
    private final RigidBodyTransform sensorToWorldForHeightMap = new RigidBodyTransform();
    private final RigidBodyTransform sensorToGroundForHeightMap = new RigidBodyTransform();
    private final RigidBodyTransform groundToWorldForHeightMap = new RigidBodyTransform();
-   private final BytedecoImage heightMapBytedecoImage;
+   private final GpuMat heightMapImage;
+   private final Mat heightmapMat;
```
Comment (TomaszTB) [inline ihmc-high-level-behaviors/src/main/java/us/ihmc/perception/RapidHeightMapManager.java]: These seem to represent the depth image from a sensor, not really the height map. I would name them something like `deviceDepthImage` and `hostDepthImage`

File: ihmc-high-level-behaviors/src/main/java/us/ihmc/perception/RapidHeightMapManager.java [hunk 2]
```diff
@@ -70,11 +73,15 @@ public void update(Mat latestDepthImage,
                       ReferenceFrame d455SensorFrame,
                       ReferenceFrame d455ZUpSensorFrame,
                       ROS2PublishSubscribeAPI ros2)
+
+
```
Comment (TomaszTB) [inline ihmc-high-level-behaviors/src/main/java/us/ihmc/perception/RapidHeightMapManager.java]: Unwanted newlines?

File: ihmc-perception/src/main/java/us/ihmc/perception/cuda/CUDAProgram.java [hunk 3]
```diff
@@ -28,8 +28,8 @@
 public class CUDAProgram implements AutoCloseable
 {
    private static final String[] DEFAULT_OPTIONS = {"-arch=" + getComputeVersion(),  // Target fairly recent GPU architecture
-                                                    "--dopt=on",                     // Optimize code
-                                                    "-G"};                           // More code optimization
+                                                    /*"--dopt=on",                     // Optimize code
+                                                    "-G"*/};                           // More code optimization
```
Comment (TomaszTB) [inline ihmc-perception/src/main/java/us/ihmc/perception/cuda/CUDAProgram.java]: You should check whether it still works with these options enabled. Ideally we should be able to have these enabled

File: ihmc-perception/src/main/java/us/ihmc/perception/cuda/CUDAStreamManager.java [hunk 4]
```diff
@@ -38,6 +38,7 @@ public static synchronized CUstream_st getStream()
       return streams.get(streamsGotten++ % streams.size()).get();
    }
 
+
```
Comment (TomaszTB) [inline ihmc-perception/src/main/java/us/ihmc/perception/cuda/CUDAStreamManager.java]: Remove newline
```suggestion
```

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractor.java
Comment (TomaszTB) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractor.java:1]: You should reset this file to develop (sorry, I'm the one who messed it up haha)

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCuda.java [hunk 5]
```diff
@@ -0,0 +1,543 @@
...
+import java.net.URL;
+import java.nio.ByteBuffer;
+import java.nio.file.Path;
+import java.util.Objects;
+
+import static org.bytedeco.cuda.global.cudart.*;
+
+import us.ihmc.perception.camera.CameraIntrinsics;
+
+public class RapidHeightMapExtractorCuda
```
Comment (TomaszTB) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCuda.java]: Capitalize "CUDA" to be consistent with rest of our code. Maybe even move it to the front of the name? Not sure about that though
```suggestion
public class RapidHeightMapExtractorCUDA
```

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCuda.java [hunk 6]
```diff
@@ -0,0 +1,543 @@
...
+         devicePointer.close();
+         System.out.println("Deallocated device pointer.");
+      }
+   }
+
+   public void initialize() throws URISyntaxException
+
+   {
+      stream = new CUstream_st();
+      cudart.cudaStreamCreate(stream);
```
Comment (TomaszTB) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCuda.java]: You should use the `CUDAStreamManager` instead (the 2D CUDA example show how it's used).

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCuda.java [hunk 7]
```diff
@@ -0,0 +1,543 @@
...
+
+      //Allocate memory on the GPU for each of the transforms and images
+      //This step involves allocating CUDA memory asynchronously, and it's important to check for allocation errors
+      CUDATools.mallocAsync(groundToSensorTransformDevicePointer, groundToSensorTransformArray.length, stream);
+      CUDATools.mallocAsync(sensorToGroundTransformDevicePointer, sensorToGroundTransformArray.length, stream);
+      CUDATools.mallocAsync(worldToGroundTransformDevicePointer, worldToGroundTransformArray.length, stream);
+      CUDATools.mallocAsync(groundToWorldTransformDevicePointer, groundToWorldTransformArray.length, stream);
+      CUDATools.mallocAsync(parametersDevicePointer, paramsArray.length, stream);
+
+      cudaStreamSynchronize(stream);
```
Comment (TomaszTB) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCuda.java]: I don't think this synchronize is needed

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCuda.java [hunk 8]
```diff
@@ -0,0 +1,543 @@
...
+
+      //Copy the data from host memory to device memory asynchronously
+      // This ensures the device has the latest data available for kernel processing
+      CUDATools.memcpyAsync(groundToSensorTransformDevicePointer, groundToSensorTransformHostPointer, groundToSensorTransformArray.length, stream);
+      CUDATools.memcpyAsync(sensorToGroundTransformDevicePointer, sensorToGroundTransformHostPointer, sensorToGroundTransformArray.length, stream);
+      CUDATools.memcpyAsync(worldToGroundTransformDevicePointer, worldToGroundTransformHostPointer, worldToGroundTransformArray.length, stream);
+      CUDATools.memcpyAsync(groundToWorldTransformDevicePointer, groundToWorldTransformHostPointer, groundToWorldTransformArray.length, stream);
+      CUDATools.memcpyAsync(parametersDevicePointer, parametersHostPointer, paramsArray.length, stream);
+
+      cudaStreamSynchronize(stream);
```
Comment (TomaszTB) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCuda.java]: This synchronize isn't needed either

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCuda.java [hunk 9]
```diff
@@ -0,0 +1,543 @@
...
+      registerKernel.withPointer(globalHeightMapImage.data()).withLong(globalHeightMapImage.step());
+      registerKernel.withPointer(parametersDevicePointer);
+      registerKernel.withPointer(worldToGroundTransformDevicePointer);
+      registerKernel.withPointer(sensorToGroundTransformDevicePointer);
+
+      croppingKernel.withPointer(globalHeightMapImage.data()).withLong(globalHeightMapImage.step());
+      croppingKernel.withPointer(sensorCroppedHeightMapImage.data()).withLong(sensorCroppedHeightMapImage.step());
+      croppingKernel.withPointer(parametersDevicePointer);
+      croppingKernel.withInt(heightMapParameters.getCropWindowSize());
+      cudaStreamSynchronize(stream);
```
Comment (TomaszTB) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCuda.java]: This isn't needed

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCuda.java [hunk 10]
```diff
@@ -0,0 +1,543 @@
...
+
+      preprocessKernel.run(stream, gridSizeKernel0, blockSize, 0);
+      updateKernel.run(stream, gridSizeKernel1, blockSize, 0);
+      cudaStreamSynchronize(stream);
```
Comment (TomaszTB) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCuda.java]: This shouldn't be needed (I'm not 100% sure on this, but I'd be very surprised if it was needed)

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCuda.java [hunk 11]
```diff
@@ -0,0 +1,543 @@
...
+      registerKernel.run(stream, gridSizeKernel2, blockSize, 0);
+
+      cudaStreamSynchronize(stream);
```
Comment (TomaszTB) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCuda.java]: This isn't needed

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCuda.java [hunk 12]
```diff
@@ -0,0 +1,543 @@
...
+      //Update the terrain map data with the new results
+      terrainMapData.setSensorOrigin(groundToWorldTransform.getTranslationX(), groundToWorldTransform.getTranslationY());
+
+      croppingKernel.run(stream, gridSizeKernel3, blockSize, 0);
+
+      cudaStreamSynchronize(stream);
+
+      Mat finalCroppedHeightMap = new Mat();  // Assuming the height map is 201x201
+      sensorCroppedHeightMapImage.download(finalCroppedHeightMap);  // Download the image from the GPU to the Mat object
+      cudaStreamSynchronize(stream);
```
Comment (TomaszTB) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCuda.java]: The GpuMat's download function is synchronous so you don't need to synchronize after it (we could use the asynchonous download function, but we haven't figured out how to do that yet)

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCuda.java [hunk 13]
```diff
@@ -0,0 +1,543 @@
...
+      System.out.println("here destroy");
+
+      heightMapCUDAProgram.close();
+      updateKernel.close();
+      registerKernel.close();
+      croppingKernel.close();
+
+      // At the end we have to destroy the stream to release the memory
+      cudart.cudaStreamDestroy(stream);
+      stream.close();
```
Comment (TomaszTB) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCuda.java]: Use `CUDAStreamManager` here too

File: ihmc-perception/src/main/java/us/ihmc/perception/tools/PerceptionDebugTools.java [hunk 14]
```diff
@@ -349,6 +349,8 @@ public static void displayHeightMap(String tag, Mat image, int delay, float scal
       display(tag, finalDisplayDepth, delay, finalDisplayDepth.rows());
    }
 
+
+
```
Comment (TomaszTB) [inline ihmc-perception/src/main/java/us/ihmc/perception/tools/PerceptionDebugTools.java]: Remove newlines
```suggestion
```

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCuda.java
Comment (TomaszTB) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCuda.java:1]: You should go through this file and do some cleanup (some unused variables, import statements, random newlines, etc). Ctrl + Alt + O to optimize imports, and Ctrl + Alt + L for the autoformatter!

File: ihmc-perception/src/main/resources/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractor.cu
Comment (TomaszTB) [inline ihmc-perception/src/main/resources/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractor.cu:1]: This file also needs some cleanup

File: ihmc-high-level-behaviors/src/test/java/us/ihmc/rdx/perception/RDXRapidHeightMapExtractorCudaDemo.java [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+import us.ihmc.rdx.ui.RDXBaseUI;
+import us.ihmc.rdx.ui.gizmo.RDXPose3DGizmo;
+import us.ihmc.rdx.ui.graphics.RDXHeightMapGraphicNew;
+import us.ihmc.robotics.referenceFrames.ZUpFrame;
+import us.ihmc.sensorProcessing.heightMap.HeightMapData;
+import us.ihmc.sensorProcessing.heightMap.HeightMapMessageTools;
+
+import java.util.List;
+
+public class RDXRapidHeightMapExtractorCudaDemo
```
Comment (TomaszTB) [inline ihmc-high-level-behaviors/src/test/java/us/ihmc/rdx/perception/RDXRapidHeightMapExtractorCudaDemo.java]:
```suggestion
public class RDXRapidHeightMapExtractorCUDADemo
```
Or whatever change to match the name change of the extractor

File: ihmc-high-level-behaviors/src/test/java/us/ihmc/rdx/perception/RDXRapidHeightMapExtractorCudaDemo.java
Comment (TomaszTB) [inline ihmc-high-level-behaviors/src/test/java/us/ihmc/rdx/perception/RDXRapidHeightMapExtractorCudaDemo.java:1]: This file could use some cleanup. Also it doesn't work for me :( Something about an index out of bounds in the `RDXHeightMapGraphicNew` class

**Review by TomaszTB — CHANGES_REQUESTED** [review]
I'm so happy with this progress. Left a bunch of comments about cleanup and whatnot, but functionality looks good... except for the demo which doesn't work for me :(

File: ihmc-perception/src/main/resources/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractor.cu [hunk 16]
```diff
@@ -0,0 +1,580 @@
...
+
+
+extern "C" __global__ void preprocessImageKernel(unsigned short *in, size_t pitchIn, unsigned short *out, size_t pitchOut)
+{
+    // Thread indices for output image
+    int xIndex = blockIdx.x * blockDim.x + threadIdx.x; // Output dimension x (new cols)
+    int yIndex = blockIdx.y * blockDim.y + threadIdx.y; // Output dimension y (new rows)
+
+    // Ensure within bounds of the output image (1280 rows and 720 columns)
+    if (xIndex >= 1280 || yIndex >= 720) {
```
Comment (ds58) [inline ihmc-perception/src/main/resources/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractor.cu]: I think this hard coded resolution needs to be passed in. Maybe add a TODO somewhere about it if we don't cover it in this PR

File: ihmc-high-level-behaviors/src/main/java/us/ihmc/perception/RapidHeightMapManager.java [hunk 17]
```diff
@@ -27,101 +28,97 @@
...
+    private final FramePose3D cameraPoseForHeightMap = new FramePose3D();
+    private final RigidBodyTransform sensorToWorldForHeightMap = new RigidBodyTransform();
+    private final RigidBodyTransform sensorToGroundForHeightMap = new RigidBodyTransform();
+    private final RigidBodyTransform groundToWorldForHeightMap = new RigidBodyTransform();
+    private final GpuMat deviceDepthImage;
+    private final Mat hostDepthImage;
+    private final Notification resetHeightMapRequested = new Notification();
+    private final BytePointer compressedCroppedHeightMapPointer = new BytePointer();
+
+    public RapidHeightMapManager(OpenCLManager openCLManager,
```
Comment (ds58) [inline ihmc-high-level-behaviors/src/main/java/us/ihmc/perception/RapidHeightMapManager.java]: We can remove `openCLManager` from the parameters, right?

Comment (LaughingCommit) [reply ihmc-high-level-behaviors/src/main/java/us/ihmc/perception/RapidHeightMapManager.java]: Other Instaces of the class use it, ill have to change it in otherplaces.

File: ihmc-perception/src/main/java/us/ihmc/perception/tools/PerceptionDebugTools.java
Comment (PotatoPeeler3000) [inline ihmc-perception/src/main/java/us/ihmc/perception/tools/PerceptionDebugTools.java:1]: @LaughingCommit This file can be reverted to develop, only formatting change

File: ihmc-perception/src/test/java/us/ihmc/perception/cuda/CUDAProgramTest.java
Comment (PotatoPeeler3000) [inline ihmc-perception/src/test/java/us/ihmc/perception/cuda/CUDAProgramTest.java:1]: @LaughingCommit This file can be reverted to develop, only formatting change.

**Review by PotatoPeeler3000 — COMMENTED** [review]
These are the files I would revert back to develop

**Review by TomaszTB — DISMISSED** [review]
I think you ran the default IntelliJ formatter on a few files :laughing:. You should undo that, and just keep whatever changes you made to the files (if you didn't make any changes, you could just reset the file to develop)

**Review by PotatoPeeler3000 — DISMISSED** [review]

**Review by PotatoPeeler3000 — APPROVED** [review]
Once the flying point commits are removed (to a different branch) this seems great!

File: ihmc-perception/src/main/java/us/ihmc/perception/filters/FlyingPointsFilter.java
Comment (ds58) [inline ihmc-perception/src/main/java/us/ihmc/perception/filters/FlyingPointsFilter.java:1]: I think we should call this CUDAFlyingPointsFilter

File: ihmc-perception/src/main/java/us/ihmc/perception/filters/CUDAFlyingPointsFilter.java [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+import org.bytedeco.opencv.global.opencv_core;
+import org.bytedeco.opencv.opencv_core.GpuMat;
+import us.ihmc.perception.cuda.CUDAKernel;
+import us.ihmc.perception.cuda.CUDAProgram;
+
+import java.net.URL;
+
+import us.ihmc.perception.cuda.CUDAStreamManager;
+
+
```
Comment (ds58) [inline ihmc-perception/src/main/java/us/ihmc/perception/filters/CUDAFlyingPointsFilter.java]: Extra newline

File: ihmc-perception/src/main/java/us/ihmc/perception/filters/CUDAFlyingPointsFilter.java [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+public class CUDAFlyingPointsFilter
```
Comment (ds58) [inline ihmc-perception/src/main/java/us/ihmc/perception/filters/CUDAFlyingPointsFilter.java]: I think this class needs some javadoc on how it works and what the filter does

File: ihmc-perception/src/main/resources/CUDA/flyingPointFilterKernel.cu
Comment (ds58) [inline ihmc-perception/src/main/resources/CUDA/flyingPointFilterKernel.cu:1]: the name should stick to capitalization convention, so `FlyingPointFilterKernel.cu`

File: ihmc-perception/src/main/java/us/ihmc/perception/filters/CUDAFlyingPointsFilter.java [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+{
+   private static final int BLOCK_SIZE_XY = 16;
+   
+   private final CUDAKernel flyingPointFilterKernel;
+   private final CUDAProgram flyingPointFilterCUDAProgram;
+   private final CUstream_st stream;
+
+   public CUDAFlyingPointsFilter()
+   {
+      stream = CUDAStreamManager.getStream();
```
Comment (ds58) [inline ihmc-perception/src/main/java/us/ihmc/perception/filters/CUDAFlyingPointsFilter.java]: You might want to pass stream into the applyFilter method instead of storing it as a field

File: ihmc-perception/src/main/java/us/ihmc/perception/filters/CUDAFlyingPointsFilter.java
Comment (ds58) [inline ihmc-perception/src/main/java/us/ihmc/perception/filters/CUDAFlyingPointsFilter.java:1]: I really think this filter should be a separate PR (perhaps debatable) and I think we needs some kind of unit test for it

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java [hunk 21]
```diff
@@ -0,0 +1,529 @@
...
+      {
+         devicePointer.close();
+         System.out.println("Deallocated device pointer.");
+      }
+   }
+
+   public void initialize() throws URISyntaxException
+   {
+      stream = new CUstream_st();
+      cudart.cudaStreamCreate(stream);
```
Comment (ds58) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java]: How come we do this instead of using the CUDAStreamManager?

File: ihmc-perception/src/main/resources/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractor.cu
Comment (ds58) [inline ihmc-perception/src/main/resources/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractor.cu:1]: I think we should put this in the same directory as the flying point filter. Just in the resources/CUDA directory

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java [hunk 22]
```diff
@@ -0,0 +1,529 @@
...
+
+      terrainCostImage = new GpuMat(globalCellsPerAxis, globalCellsPerAxis, opencv_core.CV_8UC1);
+
+      contactMapImage = new GpuMat(globalCellsPerAxis, globalCellsPerAxis, opencv_core.CV_8UC1);
+
+      sensorCroppedHeightMapImage = new GpuMat(heightMapParameters.getCropWindowSize(), heightMapParameters.getCropWindowSize(), opencv_core.CV_16UC1);
+
+      sensorCroppedTerrainCostImage = new GpuMat(heightMapParameters.getCropWindowSize(), heightMapParameters.getCropWindowSize(), opencv_core.CV_8UC1);
+
+      sensorCroppedContactMapImage = new GpuMat(heightMapParameters.getCropWindowSize(), heightMapParameters.getCropWindowSize(), opencv_core.CV_8UC1);
```
Comment (ds58) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java]: I think it would be better if we didn't have extra new lines between each of these

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java [hunk 23]
```diff
@@ -0,0 +1,529 @@
...
+      //        inputDepthImage.download(finalCroppedHeightMap);
+      //        Rect roi = new Rect(0, 0, 151, 151);
+      //        Mat croppedMat = new Mat(finalCroppedHeightMap, roi);
+      //        terrainMapData.setHeightMap(croppedMat);
+
+   }
+
+   public void destroy()
+   {
+      System.out.println("here destroy");
```
Comment (ds58) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java]: Need to remove this print or print something a bit more useful

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java [hunk 24]
```diff
@@ -0,0 +1,529 @@
...
+                                 (float) parameters.getSteppingCosineThreshold(),
+                                 (float) parameters.getSteppingContactThreshold(),
+                                 (float) parameters.getContactWindowSize(),
+                                 (float) parameters.getSpatialAlpha(),
+                                 (float) parameters.getSearchSkipSize(),
+                                 (float) parameters.getVerticalSearchSize(),
+                                 (float) parameters.getVerticalSearchResolution(),
+                                 (float) parameters.getFastSearchSize()};
+
+      initialized = true;
```
Comment (ds58) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java]: Is this actually used? I couldn't find any usages of initialized or isInitialized()

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java [hunk 25]
```diff
@@ -0,0 +1,529 @@
...
+   private GpuMat globalHeightMapImage;
+   private GpuMat globalHeightVarianceImage;
+   private GpuMat terrainCostImage;
+   private GpuMat contactMapImage;
+   private GpuMat sensorCroppedHeightMapImage;
+   private GpuMat sensorCroppedTerrainCostImage;
+   private GpuMat sensorCroppedContactMapImage;
+
+   private boolean initialized = false;
+   private int mode = 1; // 0 -> Ouster, 1 -> Realsense
```
Comment (ds58) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java:53]: Doesn't 1 mean pinhole in general? Like it should work with any generic depth sensor, right?

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java [hunk 26]
```diff
@@ -0,0 +1,529 @@
...
+   private int cropCenterIndex;
+   private int globalCellsPerAxis;
+   private CameraIntrinsics cameraIntrinsics;
+
+   private TerrainMapData terrainMapData;
+   private static HeightMapParameters heightMapParameters = new HeightMapParameters("GPU");
+   private final SideDependentList<ReferenceFrame> footSoleFrames = new SideDependentList<>();
+
+   private Rect cropWindowRectangle;
+   private static final boolean computeSteppability = true;
```
Comment (ds58) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java]: Since this is static, you should move it up above everything else

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java [hunk 27]
```diff
@@ -0,0 +1,529 @@
...
+   private float gridOffsetX;
+   private int centerIndex;
+   private int localCellsPerAxis;
+   private int globalCenterIndex;
+   private int cropCenterIndex;
+   private int globalCellsPerAxis;
+   private CameraIntrinsics cameraIntrinsics;
+
+   private TerrainMapData terrainMapData;
+   private static HeightMapParameters heightMapParameters = new HeightMapParameters("GPU");
```
Comment (ds58) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java]: Since this is static, you should move it up above everything else

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java [hunk 28]
```diff
@@ -0,0 +1,529 @@
...
+   private CUDAKernel croppingKernel;
+   private CUDAKernel preprocessKernel;
+
+   private dim3 blockSize;
+   private dim3 gridSizeKernel0;
+   private dim3 gridSizeKernel1;
+   private dim3 gridSizeKernel2;
+   private dim3 gridSizeKernel3;
+
+   float[] paramsArray;
```
Comment (ds58) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java]: I think this needs to be private

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java [hunk 29]
```diff
@@ -0,0 +1,529 @@
...
+
+   private boolean processing = false;
```
Comment (ds58) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java]: Is this used?

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java [hunk 30]
```diff
@@ -0,0 +1,529 @@
...
+
+   public void setModePerspective()
+   {
+      mode = 1;
+   }
+
+   public RapidHeightMapExtractorCUDA(ReferenceFrame leftFootSoleFrame, ReferenceFrame rightFootSoleFrame)
+   {
+      footSoleFrames.put(RobotSide.LEFT, leftFootSoleFrame);
+      footSoleFrames.put(RobotSide.RIGHT, rightFootSoleFrame);
```
Comment (ds58) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java:118]: Is it possible not to use RobotSide or reference any part of a robot in this? That way the height map extractor is very generic and can be reused without having to be specific to a humanoid robot.

File: ihmc-perception/src/main/java/us/ihmc/sensors/RealsenseColorDepthImagePublisher.java
Comment (calvertdw) [inline ihmc-perception/src/main/java/us/ihmc/sensors/RealsenseColorDepthImagePublisher.java:1]: Remember to revert these changes

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java (hunk 25 above)
Comment (LaughingCommit) [reply ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java:53]: The way it's calculated back_project_perspective/back_project_spherical is different for ouster and different for the realsense, how we find the center and calculate yaw and pitch index

File: ihmc-high-level-behaviors/src/main/java/us/ihmc/perception/StandAloneRealsenseProcess.java [hunk 31]
```diff
@@ -99,8 +98,6 @@ private void updateHeightMap()
                                  realsenseZUpFrameSupplier.get(),
                                  ros2Helper);
 
-//         PerceptionDebugTools.printHeightMap("Your mom", latestHeightMapData, 10);
```
Comment (TomaszTB) [inline ihmc-high-level-behaviors/src/main/java/us/ihmc/perception/StandAloneRealsenseProcess.java:102]: @PotatoPeeler3000 Is this your code being cleaned up :rofl:

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java (hunk 30 above)
Comment (TomaszTB) [reply ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java:118]: Yeah, it'd be nice if we could run the height map without a robot, especially for testing. Definitely a change that should be made in the future, but I don't think it's necessary for this PR.

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java [hunk 32]
```diff
@@ -0,0 +1,499 @@
...
+
+      String updateKernelName = "heightMapUpdateKernel";
+      String registerKernelName = "heightMapRegistrationKernel";
+      String croppingKernelName = "croppingKernel";
+      String preprocessKernelName = "preprocessImageKernel";
+
+      updateKernel = heightMapCUDAProgram.loadKernel(updateKernelName);
+      registerKernel = heightMapCUDAProgram.loadKernel(registerKernelName);
+      croppingKernel = heightMapCUDAProgram.loadKernel(croppingKernelName);
+      preprocessKernel = heightMapCUDAProgram.loadKernel(preprocessKernelName);
```
Comment (TomaszTB) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java]: I say we shouldn't load the kernel if it's not being used anywhere

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java [hunk 33]
```diff
@@ -0,0 +1,499 @@
...
+   {
+      if (hostPointer != null)
+      {
+         hostPointer.close();
+         System.out.println("Deallocated host pointer.");
+      }
+
+      if (devicePointer != null)
+      {
+         devicePointer.close();
```
Comment (TomaszTB) [inline ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java:171]: If this is a CUDA pointer, you need to call `cudaFree` or `cudaFreeAsync` to actually deallocate the GPU memory

**Review by TomaszTB — CHANGES_REQUESTED** [review]
Just a couple of things

File: ihmc-high-level-behaviors/src/test/resources/configurations/RDXRapidHeightMapExtractorCudaDemo/ImGuiPanels.json
Comment (ds58) [inline ihmc-high-level-behaviors/src/test/resources/configurations/RDXRapidHeightMapExtractorCudaDemo/ImGuiPanels.json:1]: This file needs to be renamed since the associated class was renamed

File: ihmc-high-level-behaviors/src/test/resources/configurations/RDXRapidHeightMapExtractorCudaDemo/ImGuiSettings.ini
Comment (ds58) [inline ihmc-high-level-behaviors/src/test/resources/configurations/RDXRapidHeightMapExtractorCudaDemo/ImGuiSettings.ini:1]: This file needs to be renamed since the associated class was renamed

File: ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java (hunk 25 above)
Comment (ds58) [reply ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java:53]: I meant that 1 isn't specific for a realsense. You can use it with any pinhole/fisheye style sensor

**Review by TomaszTB — APPROVED** [review]
Seems good enough to me. Definitely should be improved further in the future

**Review by ds58 — APPROVED** [review]
